sui_core/src/consensus_adapter.rs

```diff
@@ -129,91 +129,91 @@ impl ConsensusAdapter {
 /// notify the called when they are sequenced.
 pub struct ConsensusListener {
-    /// The authority state to execute shared-object transactions.
-    state: Arc<AuthorityState>,
     /// Receive messages input to the consensus.
-    rx_consensus_input: Receiver<ConsensusInput>,
+    rx_consensus_input: Receiver<ConsensusListenerMessage>,
     /// Receive consensus outputs.
     rx_consensus_output: Receiver<ConsensusOutput>,
+    /// The maximum number of pending replies.
+    max_pending_transactions: usize,
     /// Keep a map of all consensus inputs that are currently being sequenced.
     pending: HashMap<ConsensusTransactionDigest, Vec<Replier>>,
 }
 
 impl ConsensusListener {
     /// Spawn a new consensus adapter in a dedicated tokio task.
     pub fn spawn(
-        state: Arc<AuthorityState>,
-        rx_consensus_input: Receiver<ConsensusInput>,
+        rx_consensus_input: Receiver<ConsensusListenerMessage>,
         rx_consensus_output: Receiver<ConsensusOutput>,
+        max_pending_transactions: usize,
     ) -> JoinHandle<()> {
         tokio::spawn(async move {
             Self {
-                state,
                 rx_consensus_input,
                 rx_consensus_output,
-                pending: HashMap::new(),
+                max_pending_transactions,
+                pending: HashMap::with_capacity(2 * max_pending_transactions),
             }
             .run()
             .await
         })
     }
 
-    /// Execute sequenced transactions and notify the end-user (if any).
-    async fn process_consensus_output(&mut self, output: ConsensusOutput) {
-        let (result, serialized) = output;
-
-        // Execute the transaction (if the consensus successfully sequenced it).
-        let outcome = match result {
-            Ok(()) => {
-                let message =
-                    bincode::deserialize(&serialized).expect("Failed to deserialize consensus tx");
-                let ConsensusTransaction::UserTransaction(certificate) = message;
-                let confirmation_transaction = ConfirmationTransaction { certificate };
-                self.state
-                    .handle_confirmation_transaction(confirmation_transaction)
-                    .await
-            }
-            Err(e) => Err(SuiError::from(e)),
-        };
-
-        // Notify the caller that the transaction has been sequenced (if there is a caller).
-        let digest = Self::hash(&serialized);
-        if let Some(repliers) = self.pending.remove(&digest) {
-            for replier in repliers {
-                if replier.send(outcome.clone()).is_err() {
-                    debug!("No replier to listen to consensus output {digest}");
-                }
-            }
-        }
-    }
-
     /// Main loop receiving messages input to consensus and notifying the caller once the inputs
     /// are sequenced (of if an error happened).
     async fn run(&mut self) {
```

Comment:
Key question: why is all this happening through message passing and separate tasks instead of the task that interacts with consensus keeping the hashmap and updating when transactions go in, and when transactions go out? Are there more than 1 threads injecting and consuming messages to/from consensus?